components/brave_component_updater/browser/component_contents_verifier.h

```diff
@@ -0,0 +1,79 @@
+// Copyright (c) 2024 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#ifndef BRAVE_COMPONENTS_BRAVE_COMPONENT_UPDATER_BROWSER_COMPONENT_CONTENTS_VERIFIER_H_
+#define BRAVE_COMPONENTS_BRAVE_COMPONENT_UPDATER_BROWSER_COMPONENT_CONTENTS_VERIFIER_H_
+
+#include <string>
+#include <vector>
+
+#include "base/files/file_path.h"
+#include "base/functional/callback.h"
+#include "base/memory/ref_counted.h"
+#include "base/types/expected.h"
+
+namespace base {
+template <typename T>
+class NoDestructor;
+}  // namespace base
+
+namespace brave_component_updater {
+
+// Use on MAY_BLOCK sequence.
+class ComponentContentsAccessor
+    : public base::RefCountedThreadSafe<ComponentContentsAccessor> {
+ public:
+  enum class Error {
+    kFileReadFailed,
+    kInvalidSignature,
+  };
+
+  virtual const base::FilePath& GetComponentRoot() const = 0;
+  virtual bool IsComponentSignatureValid() const = 0;
+  virtual void IgnoreInvalidSignature(bool ignore) = 0;
+  virtual bool VerifyContents(const base::FilePath& relative_path,
+                              base::span<const uint8_t> contents) = 0;
+
+  base::expected<std::string, Error> ReadFileToString(
+      const base::FilePath& relative_path);
+  base::expected<std::vector<uint8_t>, Error> ReadFileToBytes(
+      const base::FilePath& relative_path);
+
+  std::string GetFileAsString(const base::FilePath& relative_path,
+                              const std::string& default_value);
+  std::vector<uint8_t> GetFileAsBytes(
+      const base::FilePath& relative_path,
+      const std::vector<uint8_t>& default_value);
+
+ protected:
+  friend class base::RefCountedThreadSafe<ComponentContentsAccessor>;
+  virtual ~ComponentContentsAccessor() = default;
+};
+
+class ComponentContentsVerifier {
+ public:
+  using ComponentContentsAccessorFactory =
+      base::RepeatingCallback<scoped_refptr<ComponentContentsAccessor>(
+          const base::FilePath& component_root)>;
+
+  static void Setup(ComponentContentsAccessorFactory cca_factory);
+  static ComponentContentsVerifier* GetInstance();
+
+  // Must be called on the MAY_BLOCK sequence.
+  scoped_refptr<ComponentContentsAccessor> GetContentsAccessor(
+      const base::FilePath& component_root);
+
+ private:
+  friend base::NoDestructor<ComponentContentsVerifier>;
```

Comment:
I'm not quite persuaded having a `NoDestructor` class here is the correct mechanism. This class is here so `GetContentsAccessor` can instantiate one of the two derived `ComponentContentsAccessor` classes. We have `ComponentContentsAccessorFactory` as a `RepeatingCallback`, but we also wrap `GetContentAccessor` in another repeating callback. I feel this is getting a very roundabout way to just instantiating a class. I'm particularly weary of this `NoDestructor` use here as it represents a global state for a callable.

Reply:
The same idea as `BraveOnDemandUpdater`. We need to provide objects that implemented in `browser` to the 'components`.

Reply:
This just doesn't feel truly a `NoDestructor` case. Take a look for instance at this example in upstream https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/performance_manager/metrics/metrics_provider_desktop.cc;l=32.

I feel this object should be constructed, and destructed. We already have a starting point for `ComponentContentsAccessor`, and I feel it should be the case that there's also an end point to it. We are calling `component_updater::SetupComponentContentsVerifier()` in `ChromeBrowserMainParts::PreMainMessageLoopRun()`. Is it possible for us to use a different browser event to dispose of this? If that was possible, we could be storing the pointer internally, and using something like`ComponentContentsAccessor::GetInstance()`, which removes some of this interaction.

I'm saying this because these `NoDestructor` objects are pretty much globals, and they do cause things to

Reply:
I think for this to be safe as `NoDestructor`, the factory should only create some helper that wraps `extensions::*` stuff unconditionally.